cardinal/search/composedsearch.go

```diff
@@ -0,0 +1,269 @@
+package search
+
+import (
+	"errors"
+
+	"github.com/rotisserie/eris"
+
+	"pkg.world.dev/world-engine/cardinal/search/filter"
+	"pkg.world.dev/world-engine/cardinal/types"
+	"pkg.world.dev/world-engine/cardinal/types/engine"
+)
+
+type OrSearch struct {
+	searches []Searchable
+}
+
+type AndSearch struct {
+	searches []Searchable
+}
+
+type NotSearch struct {
+	search Searchable
+}
+
+func (orSearch *OrSearch) evaluateSearch(eCtx engine.Context) []types.ArchetypeID {
+	acc := make([]types.ArchetypeID, 0)
+	for _, search := range orSearch.searches {
+		acc = append(acc, search.evaluateSearch(eCtx)...)
+	}
+	return acc
+}
+
+func (orSearch *OrSearch) Each(eCtx engine.Context, callback CallbackFn) error {
+	var err error
+	for _, search := range orSearch.searches {
+		err = errors.Join(err, search.Each(eCtx, callback))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (orSearch *OrSearch) collect(eCtx engine.Context) ([]types.EntityID, error) {
+	resMap := make(map[types.EntityID]bool)
+	for _, search := range orSearch.searches {
+		ids, err := search.collect(eCtx)
+		if err != nil {
+			return nil, err
+		}
+		for _, id := range ids {
+			resMap[id] = true
+		}
+	}
+	res := make([]types.EntityID, 0)
+	for id := range resMap {
```

Comment:
Looping over map keys is nondeterministic. res should be declared at the top of the function and appends should take place in the nested for loop near `resMap[id]=true`. The resMap should still be used to make sure duplicate ids are not added to the slice.

Reply:
I don't understand this request. Non-determinism doesn't have to do with anything here? 

It seems you're just requesting to re arrange the primitives to do the same thing? Please explain what you are thinking here.

Reply:
If I construct a Search that uses this OrSearch, and ends with a First(), a collection of entity IDs will put into a slice, and I'll be handed whatever entity ID happens to be at index 0. Given identical runs of the same tick, the set of items in the slice will always be the same, but the order of the items may change from run to run (because ranging over maps is non-deterministic). This means the entity ID returned from First() is nondeterministic.

My suggestion is to rearrange the existing primitives, but the end result will be a deterministic ordering of entity IDs that won't change between identical runs of the same tick.

If this explanation isn't clear, or if I'm missing something, let's hop on a call.

Reply:
This is correct behavior. Ordering isn't a factor in search [](edited-with-Axolo)

Reply:
I understand what you're saying [](edited-with-Axolo)

Reply:
First will be nondeterministic, we should absolutely never allow the user to rely on determinism here because it will be an arbitrary determinism/order with no rhyme or reason. [](edited-with-Axolo)

Reply:
If we want determinism that should be an explicit feature which isn't part of the spec. I would impliment it as some type of order [](edited-with-Axolo)

Reply:
can we just sort it by entity ID? search will need to return the exact same results given the same state.

Reply:
Yeah this is done.